phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,550 @@
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+        }
+    }
+
+    @VisibleForTesting
+    public void executeTool() {
+        try (Connection conn = DriverManager.getConnection(connectionURL, new Properties())) {
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+            Configuration conf = queryServices.getConfiguration();
+            executeTool(conn, queryServices, conf);
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+        }
+    }
+
+    private void executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing "+ operation);
+
+        for (Map.Entry<String, ArrayList<String>> entry: tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            ArrayList<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
+                LOGGER.fine("Executing "+operation+" for " + dataTableFullName);
+
+                boolean mutable = !(dataTable.isImmutableRows());
+
+                if (upgrade) {
+                    if (compaction) {
+                        setCompactionEnabled(admin, entry, false);
+                        checkAndWaitForMajorCompaction(admin, entry);
+                    }
+                    disableTable(admin, entry);
+                    modifyIndexTable(admin, indexes);
+                    if (mutable) {
+                        modifyDataTable(admin, dataTableFullName);
+                    }
+                    enableTable(admin, entry);
+                    rebuildIndexes(entry, conf);
+                    if (compaction) {
+                        setCompactionEnabled(admin, entry, true);
+                    }
+                } else {
+                    disableTable(admin, entry);
+                    if(mutable) {
+                        modifyDataTable(admin, dataTableFullName);
+                    }
+                    modifyIndexTable(admin, indexes);
+                    enableTable(admin, entry);
+                }
+            } catch (IOException | SQLException e) {
+                LOGGER.severe("Something went wrong while executing "+operation+" steps " + e);
```

Comment:
That's a good point. I'll include more safety methods.

Reply:
Since this tool is centered around changing the tableDescriptor for index and data table, one way of handling it could be to save older tableDescriptors before performing any operation and then in case of failure revert the table descriptors and some more atomic logic. @gjacoby126  how does this sound?

Reply:
As discussed offline, making sure the tool is completely idempotent to handle this scenario.